src/transports/ipc.rs

```diff
@@ -0,0 +1,185 @@
+    request: rpc::Request,
+    response_tx: oneshot::Sender<rpc::Value>,
+}
+
+async fn run_server(unix_stream: UnixStream, messages_rx: mpsc::UnboundedReceiver<TransportMessage>) -> Result<()> {
+    let (socket_reader, mut socket_writer) = unix_stream.into_split();
+    let mut pending_response_txs = BTreeMap::default();
+
+    let mut socket_reader = reader_stream(socket_reader);
+    let mut messages_rx = messages_rx.fuse();
+    let mut read_buffer = vec![];
+
+    loop {
+        tokio::select! {
+            message = messages_rx.next() => match message {
+                Some(message) => {
+                    if pending_response_txs.insert(message.id, message.response_tx).is_some() {
+                        log::warn!("Replacing a pending request with id {:?}", message.id);
+                    }
+
+                    let bytes = helpers::to_string(&message.request).into_bytes();
+                    if let Err(err) = socket_writer.write(&bytes).await {
+                        pending_response_txs.remove(&message.id);
+                        log::error!("IPC write error: {:?}", err);
+                    }
+                },
+                None => {},
+            },
+            bytes = socket_reader.next() => match bytes {
+                Some(Ok(bytes)) => {
+                    read_buffer.extend_from_slice(&bytes);
+
+                    let pos = match read_buffer.iter().rposition(|&b| b == b']' || b == b'}') {
+                        Some(pos) => pos,
+                        None => continue,
+                    };
+
+                    match helpers::to_response_from_slice(&read_buffer[..pos]) {
+                        Ok(response) => {
+                            let remaining_bytes_len = read_buffer.len() - pos;
+                            for i in 0..remaining_bytes_len {
+                                read_buffer[i] = read_buffer[pos + i];
+                            }
+                            read_buffer.truncate(remaining_bytes_len);
+
+                            let outputs = match response {
+                                rpc::Response::Single(output) => vec![output],
+                                rpc::Response::Batch(outputs) => outputs,
+                            };
+
+                            for output in outputs {
+                                let id = match &output {
+                                    rpc::Output::Success(success) => success.id.clone(),
+                                    rpc::Output::Failure(failure) => failure.id.clone(),
+                                };
+
+                                let value = match helpers::to_result_from_output(output) {
+                                    Ok(value) => value,
+                                    Err(err) => {
+                                        log::warn!("Unable to parse output into rpc::Value: {:?}", err);
+                                        continue;
+                                    },
```

Comment:
If you extract to a function this would become
```suggestion
                                let value = helpers::to_result_from_output(output)
                                    .map_err(|err| log::warn!("Unable to parse output into rpc::Value: {:?}", err))?;
```

which imho looks much nicer

Reply:
Done in 7f34638c896677321818bb0c1d5a93a2ad7ab3f4